transaction/std/src/memo/gift_code_funding.rs

```diff
@@ -0,0 +1,308 @@
+        let note = "";
+        assert_eq!(memo.funding_note().unwrap(), note);
+
+        // Check that the public key can be verified
+        assert!(memo.public_key_matches(&key));
+    }
+
+    #[test]
+    fn test_gift_code_funding_note_terminates_at_first_null() {
+        // Create note from bytes and put two nulls in it
+        let mut note_bytes = [b'6'; 8];
+        note_bytes[3] = 0;
+        note_bytes[6] = 0;
+        let note = "666";
+
+        // Create hash bytes
+        let mut rng: StdRng = SeedableRng::from_seed([1u8; 32]);
+        let key = RistrettoPublic::from_random(&mut rng);
+        let hash_bytes = tx_out_public_key_short_hash(&key);
+
+        // Create memo from hash & note bytes
+        let mut memo_bytes = [0u8; GiftCodeFundingMemo::MEMO_DATA_LEN];
+        memo_bytes[0..GiftCodeFundingMemo::HASH_DATA_LEN].copy_from_slice(&hash_bytes);
+        memo_bytes[GiftCodeFundingMemo::HASH_DATA_LEN..(GiftCodeFundingMemo::HASH_DATA_LEN + 8)]
+            .copy_from_slice(&note_bytes);
+        let memo = GiftCodeFundingMemo::from(&memo_bytes);
+
+        // Check that the hash is correctly verified
+        assert!(memo.public_key_matches(&key));
+
+        // Check that the note is extracted properly and terminated at first null
+        assert_eq!(memo.funding_note().unwrap(), note);
+    }
+
+    #[test]
+    fn test_gift_code_funding_memo_verified_with_wrong_public_key_doesnt_match() {
+        let mut rng: StdRng = SeedableRng::from_seed([1u8; 32]);
+        let note = "Cash money MeowbleCoin for Kitty";
+        let key = RistrettoPublic::from_random(&mut rng);
+        let other_key = RistrettoPublic::from_random(&mut rng);
+        let memo = GiftCodeFundingMemo::new(&key, note).unwrap();
+
+        // Check that a non-matching public key cannot be correctly verified
+        assert!(!memo.public_key_matches(&other_key));
+    }
+
+    #[test]
+    fn test_gift_code_funding_memo_created_with_notes_near_max_byte_lengths() {
+        // Create notes near max length
+        let note_len_minus_one =
+            str::from_utf8(&[b'6'; GiftCodeFundingMemo::NOTE_DATA_LEN - 1]).unwrap();
+        let note_len_exact = str::from_utf8(&[b'6'; GiftCodeFundingMemo::NOTE_DATA_LEN]).unwrap();
+        let note_len_plus_one =
+            str::from_utf8(&[b'6'; GiftCodeFundingMemo::NOTE_DATA_LEN + 1]).unwrap();
+        let mut rng: StdRng = SeedableRng::from_seed([1u8; 32]);
+        let key = RistrettoPublic::from_random(&mut rng);
+
+        // Create memos from notes
+        let memo_len_minus_one = GiftCodeFundingMemo::new(&key, note_len_minus_one).unwrap();
+        let memo_len_exact = GiftCodeFundingMemo::new(&key, note_len_exact).unwrap();
+        let memo_len_plus_one = GiftCodeFundingMemo::new(&key, note_len_plus_one);
+
+        // Check note lengths match or error on creation if note is too large
+        let _memo_err: Result<GiftCodeFundingMemo, MemoError> =
```

Comment:
Nit: underscore prefix is for unused variables

Reply:
Yes, while this is true, I used it because
1. The compiler didn't consider the usage within the corresponding macro below an actual usage
2. The memo didn't allow addition within a nested macro, and I wanted to avoid putting 61/65 directly. I could've potentially used assert_matches! but that's nightly rust and James has expressed the desire to move away from nightly rust - so this was the compromise :)

Reply:
That is quite funky. I think a potential solution is something like this:
`assert!(matches!(memo_len_plus_one, Err(MemoError::BadLength(len)) if len == GiftCodeFundingMemo::NOTE_DATA_LEN + 1)`
Would you mind trying that please?

Reply:
That failed, but found an alternate solution.